whylog/rulesbase/__init__.py

```diff
@@ -1,6 +1,144 @@
+import re
+import yaml
+from abc import ABCMeta, abstractmethod
+from collections import defaultdict
 
+class AbstractConfigAccess(object):
+    __metaclass__ = ABCMeta
 
-class WhylogBase(object):
+    @abstractmethod
+    def get_parsers(self):
+        pass
 
-    def __init__(self, **kwargs):
+    @abstractmethod
+    def get_rule(self, effect_name):
         pass
+        
+class YamlConfigAccess(AbstractConfigAccess):
+    def __init__(self, parsers_path, rules_path):
+        self._parsers_path = parsers_path
+        self._rules_path = rules_path
+        self._load_parsers()
+        self._load_rules()
+
+    def _load_config_from_file(self, path):
+        with open(path, "r") as config_file:
+            return list(yaml.load_all(config_file))
+
+    def _get_params_for_parser(self, parser_dict):
+        if 'params' in parser_dict:
+            return parser_dict.get('params')
+        else:
+            return None
+
+    def _load_parsers(self):
+        parsers = self._load_config_from_file(self._parsers_path)
+        parsers_dict = {}
+        for parser in parsers:
+            parsers_dict[parser.get('name')] = Parser(
+                parser.get('name'), 
+                parser.get('regex'), 
+                self._get_params_for_parser(parser)
+            )
+        self._parsers = parsers_dict
+
+    def _get_causes_for_rule(self, rule_dict):
+        if 'cause' in rule_dict:
+            return [rule_dict.get('cause')]
+        else:
+            return rule_dict.get('causes')
+
+    def _get_constraints_for_rule(self, rules_dict):
+        if 'constraints' in rules_dict:
+            return rules_dict.get('constraints')
+        else:
+            return None
+
+    def _get_type_for_rule(self, rules_dict):
+        if 'type' in rules_dict:
+            return rules_dict.get('type')
+        else:
+            return None
+
+    def _get_cause_constraints_for_rule(self, rules_dict):
+        if 'cause_constraints' in rules_dict:
+            return rules_dict.get('cause_constraints')
+        else:
+            return None
+
+    def _load_rules(self):
```

Comment:
almost everything can be not supplied, leading to objects of YamlConfigAccess with nearly all fields filled with None. This will load the config fine, but then the whole program will need to anticipate that there can be None values everywhere. It will be very fragile.

Instead you could use empty containers, so if there are no "constraints" in the configuration and someone asks about them, you return an empty list, which the caller then iterates over and can see that all constraints are met.